src/overloads.ts

```diff
@@ -0,0 +1,93 @@
+import {StrictEqualUsingTSInternalIdenticalToOperator, IsNever} from './utils'
+
+// prettier-ignore
+/**
+ * The simple(ish) way to get overload info from a function `F`. Recent versions of TypeScript will match any function against a generic 10-overload type, filling in slots with duplicates of the function.
+ * So, we can just match against a single type and get all the overloads.
+ *
+ * For older versions of TypeScript, we'll need to painstakingly do ten separate matches.
+ */
+export type TSPost53OverloadsInfoTuple<F> =
+  F extends {(...args: infer A1): infer R1; (...args: infer A2): infer R2; (...args: infer A3): infer R3; (...args: infer A4): infer R4; (...args: infer A5): infer R5; (...args: infer A6): infer R6; (...args: infer A7): infer R7; (...args: infer A8): infer R8; (...args: infer A9): infer R9; (...args: infer A10): infer R10}
+  ? [{parameters: A1; return: R1}, {parameters: A2; return: R2}, {parameters: A3; return: R3}, {parameters: A4; return: R4}, {parameters: A5; return: R5}, {parameters: A6; return: R6}, {parameters: A7; return: R7}, {parameters: A8; return: R8}, {parameters: A9; return: R9}, {parameters: A10; return: R10}, ]
+  : never
+
+export type BaseOverloadInfo = {parameters: unknown[]; return: unknown}
+
+/**
+ * `true` iff `T` is equivalent to `{parameters: unknown[]; return: unknown}`, which is what an overload info object looks like for a non-existent overload
+ * This is useful because older versions of TypeScript end up with 9 "useless" overloads and one real one for parameterless/generic functions
+ */
+export type IsUselessOverloadInfo<T> = StrictEqualUsingTSInternalIdenticalToOperator<T, BaseOverloadInfo>
+
+// prettier-ignore
```

Comment:
I'd rather keep them because otherwise prettier makes these types extremely long - it always puts function overloads on individual lines. Better IMO to have one type per line since there's very little interesting information in the explicit counting from 1 to 10